lib/options.ts

```diff
@@ -1,63 +1,98 @@
 
-  // Check for invalid.
-  const invalid = values.find((val) => !VALUES_OF_TYPE.has(val));
-  if (invalid) {
-    throw new Error(`Invalid ruleListColumns option: ${invalid}`);
-  }
-  if (values.length !== new Set(values).size) {
-    throw new Error('Duplicate value detected in ruleListColumns option.');
-  }
+export const NOTICE_TYPE_DEFAULT_PRESENCE_AND_ORDERING: {
+  [key in NOTICE_TYPE]: boolean;
+} = {
+  // Object keys ordered in display order.
+  // Object values indicate whether the column is displayed by default.
+  [NOTICE_TYPE.DEPRECATED]: true, // Most important.
+  [NOTICE_TYPE.CONFIGS]: true,
+  [NOTICE_TYPE.FIXABLE]: true,
+  [NOTICE_TYPE.FIXABLE_AND_HAS_SUGGESTIONS]: true, // Potentially replaces FIXABLE and HAS_SUGGESTIONS.
+  [NOTICE_TYPE.HAS_SUGGESTIONS]: true,
+  [NOTICE_TYPE.REQUIRES_TYPE_CHECKING]: true,
+  [NOTICE_TYPE.TYPE]: false,
+};
 
-  if (values.length === 0) {
-    // Use default presence and ordering.
-    values.push(
-      ...Object.entries(COLUMN_TYPE_DEFAULT_PRESENCE_AND_ORDERING)
-        .filter(([_type, enabled]) => enabled)
-        .map(([type]) => type)
-    );
-  }
-
-  return values as COLUMN_TYPE[];
+export enum OPTION_TYPE {
+  CHECK = 'check',
+  CONFIG_EMOJI = 'configEmoji',
+  IGNORE_CONFIG = 'ignoreConfig',
+  IGNORE_DEPRECATED_RULES = 'ignoreDeprecatedRules',
+  PATH_RULE_DOC = 'pathRuleDoc',
+  PATH_RULE_LIST = 'pathRuleList',
+  RULE_DOC_NOTICES = 'ruleDocNotices',
+  RULE_DOC_SECTION_EXCLUDE = 'ruleDocSectionExclude',
+  RULE_DOC_SECTION_INCLUDE = 'ruleDocSectionInclude',
+  RULE_DOC_SECTION_OPTIONS = 'ruleDocSectionOptions',
+  RULE_DOC_TITLE_FORMAT = 'ruleDocTitleFormat',
+  RULE_LIST_COLUMNS = 'ruleListColumns',
+  SPLIT_BY = 'splitBy',
+  URL_CONFIGS = 'urlConfigs',
 }
 
-/**
- * Parse the option, check for errors, and set defaults.
- */
-export function parseRuleDocNoticesOption(
-  ruleDocNotices: string | undefined
-): NOTICE_TYPE[] {
-  const values = ruleDocNotices ? ruleDocNotices.split(',') : [];
-  const VALUES_OF_TYPE = new Set(Object.values(NOTICE_TYPE).map(String));
-
-  // Check for invalid.
-  const invalid = values.find((val) => !VALUES_OF_TYPE.has(val));
-  if (invalid) {
-    throw new Error(`Invalid ruleDocNotices option: ${invalid}`);
-  }
-  if (values.length !== new Set(values).size) {
-    throw new Error('Duplicate value detected in ruleDocNotices option.');
-  }
+const DEFAULT_RULE_DOC_TITLE_FORMAT: RuleDocTitleFormat =
+  'desc-parens-prefix-name'; // Using this variable ensures this default has the correct type (not just a plain string).
 
-  if (values.length === 0) {
-    // Use default presence and ordering.
-    values.push(
-      ...Object.entries(NOTICE_TYPE_DEFAULT_PRESENCE_AND_ORDERING)
-        .filter(([_type, enabled]) => enabled)
-        .map(([type]) => type)
-    );
-  }
+// TODO: use TypeScript 4.9 satisfies keyword ([key in OPTION_TYPE]...) to ensure all options are included without losing type information.
+export const OPTION_DEFAULTS = {
```

Comment:
You can use `Record` instead. You will get type errors if you the values aren't in sync with `OPTION_TYPE`

```suggestion
export const OPTION_DEFAULTS: Record<OPTION_TYPE, string> = {
```

Reply:
Thanks for the suggestion, just tried it and it looks like it's equivalent to the type I was originally trying to use `{ [key in OPTION_TYPE]: string | string[] | boolean | undefined }` and has the same problem. While it does correctly ensure all options are included in the object, it loses type information since accessing a single option from the object now could have any of those possible types instead of what should be a known type. The new [satisfies](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-beta/#the-satisfies-operator) feature in TypeScript 4.9 looks like it will solve this.

Reply:
Can you elaborate on why you need the exact type for each of them?
It doesn't seem like it gives extra value 🤔

Reply:
It's required actually. This object stores default values for each config option. These default values are used throughout the codebase. When they have inexact types (i.e. could be one of several types), they won't be able to be used easily due to type mismatches. For example, for an option with an array default value, we won't be able to handle/iterate the default value properly if the type says it could also be a string, boolean, etc.